src/store/index.js

```diff
@@ -1,14 +1,33 @@
 import { createStore } from 'redux';
 import reducer from '../reducers';
-import { saveState } from '../utils';
+import { resetData } from '../actions';
+import getInitialState from './initial-state';
+import loadData from './load-data';
+import { saveState } from './helpers';
+import checkFontLoaded from './check-font-loaded';
+
+export default function configureStore(props) {
+  /**
+   * Dispatch an action to update the store with new pipeline data
+   * @param {Object} normalizedData Normalised state data
+   */
+  const resetStoreData = normalizedData => {
+    if (store) {
+      store.dispatch(resetData(normalizedData));
+    }
+  };
+
+  const pipelineData = loadData(props.data, resetStoreData);
```

Comment:
It seems there's a bit of a mismatch in how data is injected into the store i.e. inside configureStore it's actually loading data via loadData in configureStore but then in the reset data it's triggering an action with some data, what i'd recommend is to consolidate the mental model with data in the store e.g.

- Remove onLoadData callback from loadData as it only happens in one case of the switch and feels like a convenience callback rather than something that adding value to the flow (I still think loadData doing very different things based on whether you pass a string, array or nothing to it needs a rethink but not for now 🙂 
- Remove loadData call from inside configureStore - It should focus purely on store configuring rather than making decisions itself
- Pass in fully formed initial data into configureStore from App i.e. loadData in app

```

```

Reply:
I'd love to pair with you on this because I agree that it's a little janky how it's handled currently, but I couldn't figure out a better way to structure the code while keeping the outward behaviour as is.

Here are some stipulations/assumptions about application usage:

1. The application can use a variety of different data sources:
    - test datasets (x2), imported directly via ES6 imports
    - demo dataset, imported directly via ES6 import
    - JSON file dataset at `/public/api/nodes.json`, imported asynchronously and parsed with `d3.json`
    - randomly-generated dataset, imported directly via ES6 import and generated at runtime
    - JS dataset object literal, passed as a prop to the App component
2. The application can be run in a variety of different contexts:
    - standalone JS app in development (which can run/test any of the different available data sources)
    - as a static build (usually served via a local Python server), which imports the JSON, which is served at `/api/nodes.json`
    - as a React component, imported into another React application, and passed the data as a prop.
3. When the application is run as a React component, it should update when passed new data via the `data` prop. This should only reset the chart data - there is no need to reset the rest of the state (e.g. the stuff set in `getInitialState`) when updating the chart data.
4. When loading the data asynchronously, the state will need to be initialised immediately, then updated once the data is loaded. This update is triggered with a Promise callback.

Because it can use data from different formats and in different contexts, I want to keep it simple, so I'm just exposing a single top-level `data` prop, and the application can figure out what to do with that by evaluating the value provided. This is what `loadData` does. I could name it something else but I haven't thought of a better name.

- If you remove the `onLoadData` callback then the state won't be updated when loading the data asynchronously, which means that in the main production use-case, no data will be displayed. We can structure this differently, but ultimately the callback needs to either initialise or update the state once it's triggered.

- I can split up configureStore into multiple functions, but it's tricky because `resetStoreData` needs access to the `store` object, but it also needs to be defined before it's created. It's a bit of a Gordian Knot. I don't think splitting it makes it much better tbh!

- Passing in fully-formed data into configureStore from App is what I had it doing initially, and it felt like too much was going on in App so I wanted to move this logic into `configureStore`! I don't see how placing this logic in `configureStore` makes much difference.

Basically, I don't think I can do what you're asking me to without breaking some of the required functionality.

Reply:
See [this commit](https://github.com/quantumblacklabs/kedro-viz/pull/106/commits/504989be386d78c34888ecbe3e28255c1b0d7ddb)

Reply:
Happy to pair on this, essentially it's about separation of concerns with this function - it's doing too much (loadData + store data injection). We need to separate out the applications contexts because right now it's not obvious what happens and when it happens (and you're masking it by wrapping everything in the one method). This will become brittle over time without some structure in place. Happy to chat further and not block this PR 👍

Reply:
Thanks. Yeah tbh in refactoring all I'm doing is shifting this logic back and forth between the App component and the store function. This also works:

```
class App extends React.Component {
  constructor(props) {
    super(props);
    const pipelineData = loadData(props.data, this.resetStoreData.bind(this));
    const initialState = getInitialState(pipelineData, props);
    this.store = configureStore(initialState);
  }
 ...
}
```
but now it's back to how it was before. I started this change because I was trying to move the state initialisation out of App while allowing it access to the store (which it needs for the reset)